examples/greeter-extension/README.md

```diff
@@ -0,0 +1,160 @@
+# @loopback/example-greeter-extension
+
+This example project illustrates how to implement the
+[Extension Point/extension pattern](https://wiki.eclipse.org/FAQ_What_are_extensions_and_extension_points%3F),
+which promotes loose coupling and offers great extensibility. There are many use
+cases in LoopBack 4 that fit into design pattern. For example:
+
+- `@loopback/boot` uses `BootStrapper` that delegates to `Booters` to handle
+  different types of artifacts
+- `@loopback/rest` uses `RequestBodyParser` that finds the corresponding
+  `BodyParsers` to parse request body encoded in different media types
+
+## Overview
+
+We'll use the following scenario to walk through important steps to organize the
+`greet` service that allows extensible languages - each of them being supported
+by a `Greeter` extension.
+
+![greeters](greeters.png)
+
+Various constructs from LoopBack 4, such as `Context`, `@inject.*`, and
+`Component` are used to build the service in an extensible fashion.
+
+## Define an extension point
+
+In our scenario, we want to allow other modules to extend or customize how
+people are greeted in different languages. To achieve that, we declare the
+`greeter` extension point, which declares a contract as a TypeScript interface
+that extensions must conform to.
+
+### Define interface for extensions
+
+```ts
+/**
+ * Typically an extension point defines an interface as the contract for
+ * extensions to implement
+ */
+export interface Greeter {
+  language: string;
+  greet(name: string): string;
+}
+```
+
+### Define class for the extension point
+
+```ts
+/**
+ * An extension point for greeters that can greet in different languages
+ */
+export class GreeterExtensionPoint {
```

Comment:
I find this class name problematic. This class is not the extension point, it's a service providing multilingual `greet` function and tapping into the `greeter` extension point. The actual extension point is just a convention on how greeter implementations should be bound on the context and how the greeter service is looking them up.

If we want to better show the implementation details of an extension point, then I am proposing a slightly different design:

- `Greeter` interface remains without changes, this part is already designed wells.
- `GreeterExtensionPoint` that accepts the requested language and returns `Greeter` instance matching the language.
- `GreeterService` that provides `greet` method and uses `GreeterExtensionPoint` to obtain the greeter instance to use.

```ts
export class GreeterExtensionPoint {
  constructor(
    @inject.getter(bindingTagFilter({extensionPoint: 'greeter'}))
    private greeters: Getter<Greeter[]>,
  ) {}

  findByLanguage(language: string): Greeter | undefined {
    const greeters = await this.greeters();
    return greeters.find(g => g.language === language);
  }
}

class GreeterService
  constructor(
    @inject('extensions.greeter)
    private greeterRegistry: GreeterExtensionPoint
    // ...
  ) {}

  async greet(language: string, name: string): Promise<string> {
    const greeter = await greeterRegistry.findByLanguage(language);
    const greeting = greeter ? greeter.greet(name) : `Hello, ${name}`;
    // etc.
  }
}
```

Reply:
I feel it would be even better if we could find a way how to refactor GreeterExtensionPoint into something more generic that can be configured via `@inject` decorator. Below is the first version that comes to my mind. It's not as pretty as I would like it to be, we may need to look for a different approach.

```ts
type GreeterLocator = (lang: string) => Greeter;

class GreeterService {
  constructor(
    @inject.extensionLocator(
       'extensions.greeter', 
       (lang: string) => (ext: Greeter) => ext.language === lang)
    private findGreeterByLanguage: GreeterLocator) {}

   async greet(language: string, name: string): Promise<string> {
    const greeter = await findGreeterByLanguage(language);
    // etc.
  }
}

// resolver for extensionLocator
// the next two values will be obtained from injection metadata
const extensionPointName = // 'extensions.greeter';
const filterFactory = // (lang: string) => (ext: Greeter) => ext.language === lang);
const extensions = await new ContextView(ctx, filterByExtension(extensionPointName)).values();
return (...args) => {
  const filter = filterFactory(...args);
  return extensions.find(filter);
};
```

Maybe the complexity is not worth the benefits? IDK.

Anyways, this is probably out of scope of this pull request.

Reply:
I was experimenting generalization of extension points into a base class at https://github.com/strongloop/loopback-next/tree/extension-point. My current assessment is that the benefit is not significant enough.